### PR TITLE
Partition: add key-partition document

### PR DESCRIPTION
**Pull request by qiancai**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x master (the latest development version)
- [x] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/13223
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: partitioned-table.md [hunk 1]
```diff
@@ -722,6 +815,10 @@ You can see that the inserted record `(NULL, 'mothra')` falls into the same part
 >
 > In this case, the actual behavior of TiDB is in line with the description of this document.
 
+#### Handling of NULL with Key partitioning
```
Comment (qiancai) [inline partitioned-table.md:818]:
```suggestion
#### Handling of NULL with Key partitions
```

File: partitioned-table.md [hunk 2]
```diff
@@ -722,6 +815,10 @@ You can see that the inserted record `(NULL, 'mothra')` falls into the same part
 >
 > In this case, the actual behavior of TiDB is in line with the description of this document.
 
+#### Handling of NULL with Key partitions
+
+When partitioning tables by Key, the way of handling `NULL` value is consistent with that of Hash partitioning - if the value of the partitioning field is `NULL`, it is considered as `0`.
```
Comment (mjonss) [inline partitioned-table.md:820]: Is this verified? I would rather say that all NULLs will be placed in the same partition.

File: partitioned-table.md [hunk 3]
```diff
@@ -730,6 +827,8 @@ For `LIST` and `RANGE` partitioned tables, `REORGANIZE PARTITION` is not yet sup
 
 For `HASH` partitioned tables, `COALESCE PARTITION` and `ADD PARTITION` are not yet supported.
 
+For `Key` partitioned tables, only `ALTER TABLE ... TRUNCATE PARTITION` is supported.
+
```
Comment (mjonss) [inline partitioned-table.md:831]:
```suggestion
For `HASH` and `KEY` partitioned tables, `COALESCE PARTITION` and `ADD PARTITION` are not yet supported.

For `HASH` and `KEY` partitioned tables, only `ALTER TABLE ... TRUNCATE PARTITION` is supported.

```

File: partitioned-table.md [hunk 4]
```diff
@@ -841,6 +940,12 @@ alter table members optimize partition p0;
 ERROR 8200 (HY000): Unsupported optimize partition
 ```
 
+### Key partition management
+
+Currently, Key partitioning only supports the `ALTER TABLE ... TRUNCATE PARTITION` partition management statement.
+
+If you execute a Key partition management statement that is not yet supported, TiDB will return an error.
+
```
Comment (mjonss) [inline partitioned-table.md:948]: Is this needed?

File: partitioned-table.md [hunk 5]
```diff
@@ -1350,7 +1455,9 @@ YEARWEEK()
 
 ### Compatibility with MySQL
 
-Currently, TiDB supports Range partitioning, Range COLUMNS partitioning, List partitioning, List COLUMNS partitioning, and Hash partitioning. Other partitioning types that are available in MySQL such as key partitioning are not supported yet in TiDB.
+Currently, TiDB supports Range partitioning, Range COLUMNS partitioning, List partitioning, List COLUMNS partitioning, Hash partitioning, and Key partitioning. Other partitioning types that are available in MySQL are not supported yet in TiDB.
+
+Currently, TiDB only supports single-column partitioning for Range COLUMNS partitioning and does not support using an empty partition column list for Key partitioning.
```
Comment (mjonss) [inline partitioned-table.md:1460]:
```suggestion
Currently, TiDB does not support using an empty partition column list for Key partitioning.
```

Multi-column RANGE COLUMNS partitioning is supported, see #range-columns-partitioning

File: partitioned-table.md [hunk 6]
```diff
@@ -570,6 +657,12 @@ Since v6.4.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax
 
 - For all other statements of MySQL Linear Hash partitions, they work in TiDB the same as that in MySQL, except that the rows are distributed differently if the number of partitions is not a power of 2, which will give different results for [partition selection](#partition-selection), `TRUNCATE PARTITION`, and `EXCHANGE PARTITION`.
 
+### How TiDB handles Linear Key partitions
+
+Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.
```
[the remark this answers is not in the document]
Comment (TonsnakeLin) [reply partitioned-table.md:662]: `TiDB supports **parsing** ...` which says TiDB only supports parsing LINEAR KEY, not support  LINEAR KEY.

File: partitioned-table.md (hunk 2 above)
Comment (TonsnakeLin) [reply partitioned-table.md:820]: Yes

File: partitioned-table.md [hunk 7]
```diff
@@ -560,7 +561,93 @@ MOD(YEAR('2005-09-01'),4)
...
+    separated DATE DEFAULT '9999-12-31',
+    job_code INT,
+    store_id INT
+)
+
+PARTITION BY KEY(fname, store_id)
+PARTITIONS 4;
+```
+
+Currently, TiDB does not support creating Key partitioned tables if the partition column list specified in  `PARTITION BY KEY` is empty. For example, after you execute the following statements, TiDB will create a non-partitioned table and return a `Unsupported partition type RANGE, treat as normal table` warning.
```
[the remark this answers is not in the document]
Comment (TonsnakeLin) [reply partitioned-table.md:631]: Yes

File: partitioned-table.md [hunk 8]
```diff
@@ -10,9 +10,10 @@ This document introduces TiDB's implementation of partitioning.
...
+This section introduces the types of partitioning in TiDB. Currently, TiDB supports [Range partitioning](#range-partitioning), [Range COLUMNS partitioning](#range-columns-partitioning), [List partitioning](#list-partitioning), [List COLUMNS partitioning](#list-columns-partitioning), [Hash partitioning](#hash-partitioning), and [Key partitioning](#key-partitioning).
 
-Range partitioning, Range COLUMNS partitioning, List partitioning and List COLUMNS partitioning are used to resolve the performance issues caused by a large amount of deletions in the application, and support fast drop partition operations. Hash partitioning is used to scatter the data when there are a large amount of writes.
+- Range partitioning, Range COLUMNS partitioning, List partitioning and List COLUMNS partitioning are used to resolve the performance issues caused by a large amount of deletions in the application, and support fast drop partition operations. Hash partitioning is used to scatter the data when there are a large amount of writes.
```
Comment (qiancai) [inline partitioned-table.md:15]:
```suggestion
- Range partitioning, Range COLUMNS partitioning, List partitioning, and List COLUMNS partitioning are used to resolve the performance issues caused by a large number of deletions in the application, and support fast drop partition operations. 
```

File: partitioned-table.md [hunk 9]
```diff
@@ -10,9 +10,10 @@ This document introduces TiDB's implementation of partitioning.
...
+- Hash partitioning and Key partitioning are used to distribute data in scenarios with a large number of writes. Compared with Hash partitioning, Key partitioning supports distributing data of multiple columns and partitioning by non-integer fields.
```
Comment (qiancai) [inline partitioned-table.md:16]:
```suggestion
- Hash partitioning and Key partitioning are used to distribute data in scenarios with a large number of writes. Compared with Hash partitioning, Key partitioning supports distributing data of multiple columns and partitioning by non-integer columns.
```

File: partitioned-table.md [hunk 10]
```diff
@@ -560,7 +561,93 @@ MOD(YEAR('2005-09-01'),4)
 =  1
 ```
 
-#### How TiDB handles Linear Hash partitions
+### Key partitioning
+
+Starting from v7.0.0, TiDB supports Key partitioning. For TiDB versions earlier than v7.0.0, if you try creating a Key partitioned table, TiDB will create it as a non-partitioned table and return a warning.
+
+Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or column, while Key partitioning supports distributing data based on a column list, and the type of the partitioning column for Key partitioning is not limited to integer types. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the data distribution of the table is also different.
```
Comment (qiancai) [inline partitioned-table.md:568]:
```suggestion
Both Key partitioning and Hash partitioning can evenly distribute data into a certain number of partitions. The difference is that Hash partitioning only supports distributing data based on a specified integer expression or an integer column, while Key partitioning supports distributing data based on a column list, and partitioning columns of Key partitioning are not limited to the integer type. The Hash algorithm of TiDB for Key partitioning is different from that of MySQL, so the table data distribution is also different.
```

File: partitioned-table.md [hunk 11]
```diff
@@ -560,7 +561,93 @@ MOD(YEAR('2005-09-01'),4)
...
+
+Partitioning by Key requires you to append a `PARTITION BY Key (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, `GEOMETRY`. In addition, you might also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into.
```
Comment (qiancai) [inline partitioned-table.md:570]:
```suggestion
Partitioning by Key requires you to append a `PARTITION BY Key (columList)` clause to the `CREATE TABLE` statement. `columList` is a column list with one or more column names. The data type of each column in the list can be any type except `BLOB`, `JSON`, and `GEOMETRY` (Note that TiDB does not support the `GEOMETRY`). In addition, you might also need to append `PARTITIONS num`, where `num` is a positive integer indicating how many partitions a table is divided into.
```

File: partitioned-table.md [hunk 12]
```diff
@@ -560,7 +561,93 @@ MOD(YEAR('2005-09-01'),4)
...
+    store_id INT
+)
+
+PARTITION BY KEY(store_id)
+PARTITIONS 4;
+```
+
+If `PARTITIONS num` is not specified, the default number of partitions is 1.
+
+You can also create a Key partitioned table based on non-integer fields such as VARCHAR. For example, you can partition a table by the `fname` column:
```
Comment (qiancai) [inline partitioned-table.md:593]:
```suggestion
You can also create a Key partitioned table based on non-integer columns such as VARCHAR. For example, you can partition a table by the `fname` column:
```

File: partitioned-table.md [hunk 13]
```diff
@@ -560,7 +561,93 @@ MOD(YEAR('2005-09-01'),4)
...
+    separated DATE DEFAULT '9999-12-31',
+    job_code INT,
+    store_id INT
+)
+
+PARTITION BY KEY(fname)
+PARTITIONS 4;
+```
+
+You can also create a Key partitioned table based on multiple columns. For example, you can partition a table into 4 partitions based on `fname` and `store_id`:
```
Comment (qiancai) [inline partitioned-table.md:612]:
```suggestion
You can also create a Key partitioned table based on multiple columns. For example, you can divide a table into 4 partitions based on `fname` and `store_id`:
```

File: partitioned-table.md [hunk 14]
```diff
@@ -570,6 +657,12 @@ Since v6.4.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax
 
 - For all other statements of MySQL Linear Hash partitions, they work in TiDB the same as that in MySQL, except that the rows are distributed differently if the number of partitions is not a power of 2, which will give different results for [partition selection](#partition-selection), `TRUNCATE PARTITION`, and `EXCHANGE PARTITION`.
 
+### How TiDB handles Linear Key partitions
+
+Starting from v7.0.0, TiDB supports parsing the MySQL `PARTITION BY LINEAR HASH` syntax for Key partitioning. However, TiDB ignores the `LINEAR` keyword and uses a non-linear hash algorithm instead.
+
+Before v7.0.0, if you create a Key partitioned table, TiDB will create it as a non-partitioned table and return a warning.
```
Comment (qiancai) [inline partitioned-table.md:664]:
```suggestion
Before v7.0.0, if you try creating a Key partitioned table, TiDB will create it as a non-partitioned table and return a warning.
```

File: partitioned-table.md (hunk 2 above)
Comment (qiancai) [inline partitioned-table.md:820]:
```suggestion
For Key partitioning, the way of handling `NULL` value is consistent with that of Hash partitioning. If the value of a partitioning field is `NULL`, it is considered as `0`.
```

File: partitioned-table.md [hunk 15]
```diff
@@ -722,6 +815,10 @@ You can see that the inserted record `(NULL, 'mothra')` falls into the same part
 >
 > In this case, the actual behavior of TiDB is in line with the description of this document.
 
+#### Handling of NULL with Key partitions
```
Comment (qiancai) [inline partitioned-table.md:818]:
```suggestion
#### Handling of NULL with Key partitioning
```

File: partitioned-table.md [hunk 16]
```diff
@@ -841,6 +940,12 @@ alter table members optimize partition p0;
 ERROR 8200 (HY000): Unsupported optimize partition
 ```
 
+### Key partition management
+
+Currently, Key partitioning only supports the `ALTER TABLE ... TRUNCATE PARTITION` partition management statement.
+
+If you execute a Key partition management statement that is not yet supported, TiDB will return an error.
```
Comment (qiancai) [inline partitioned-table.md:947]:
````suggestion
If you execute a Key partition management statement that is not yet supported, TiDB will return an error.

```sql
ALTER TABLE members OPTIMIZE PARTITION p0;
```

```sql
ERROR 8200 (HY000): Unsupported optimize partition
```
````

File: partitioned-table.md (hunk 8 above)
Comment (qiancai) [inline partitioned-table.md:15]:
```suggestion
- Range partitioning, Range COLUMNS partitioning, List partitioning, and List COLUMNS partitioning are used to resolve the performance issues caused by a large number of deletions in the application, and support fast drop partition operations.
```

Comment (qiancai) [inline partitioned-table.md:15]:
```suggestion
- Range partitioning, Range COLUMNS partitioning, List partitioning, and List COLUMNS partitioning are used to resolve the performance issues caused by a large number of deletions in the application, and support fast drop partition operations.
```

**Comment by qiancai** [comment]
Close this PR because the source repo has been deleted and further edits are not allowed by GitHub. Here's the latest PR for the same changes: https://github.com/pingcap/docs/pull/12963

**Pull request closed without merging**